File: rust/src/shorthand/lexer.rs
//! Hand-written scanner: `.jfn` source text -> a flat token stream.
//!
//! Kept deliberately hand-rolled (no regex, no generator) so its behavior can
//! be mirrored byte-for-byte across the Go / Python / TypeScript ports, the
//! same way `jsonc.rs` mirrors Go's `StripJSONC`.

use serde_json::Number;

use super::error::ParseError;

/// A lexical token. String and number literals carry their *decoded* value
/// (escapes resolved), so the parser never re-decodes.
#[derive(Debug, Clone, PartialEq)]
pub enum Tok {
    Num(Number),
    Str(String),
    Ident(String),
    /// A backtick template already split into literal spans and raw hole
    /// sources. The parser recursively parses each hole's source string.
    Template(Vec<TemplatePart>),

    LParen,
    RParen,
    LBracket,
    RBracket,
    LBrace,
    RBrace,
    Comma,
    Colon,
    Dot,
    DotDotDot, // ...  (rest parameter)
    Arrow,     // ->   (cond/match arm)
    FatArrow,  // =>   (function literal)
    Amp,       // &    (function reference)

    Bang, // !
    Plus,
    PlusPlus, // ++
    Minus,
    Star,
    Slash,
    Percent,

    EqEq,  // ==
    BangEq, // !=
    Lt,
    LtEq,
    Gt,
    GtEq,

    AndAnd, // &&
    OrOr,   // ||

    Eof,
}

/// One piece of a backtick template: a literal span or a `${ ... }` hole whose
/// raw (undecoded) expression source is parsed later.
#[derive(Debug, Clone, PartialEq)]
pub enum TemplatePart {
    Lit(String),
    Hole(String),
}

/// A token plus its 1-based source position (start of the token).
#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub tok: Tok,
    pub line: usize,
    pub col: usize,
}

struct Lexer {
    chars: Vec<char>,
    i: usize,
    line: usize,
    col: usize,
}

/// Tokenize `src`, returning tokens terminated by a single [`Tok::Eof`].
pub fn lex(src: &str) -> Result<Vec<Token>, ParseError> {
    let mut lx = Lexer {
        chars: src.chars().collect(),
        i: 0,
        line: 1,
        col: 1,
    };
    lx.run()
}

impl Lexer {
    fn peek(&self) -> Option<char> {
        self.chars.get(self.i).copied()
    }

    fn peek2(&self) -> Option<char> {
        self.chars.get(self.i + 1).copied()
    }

    fn bump(&mut self) -> Option<char> {
        let c = self.chars.get(self.i).copied()?;
        self.i += 1;
        if c == '\n' {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        Some(c)
    }

    fn err(&self, msg: impl Into<String>) -> ParseError {
        ParseError::new(msg, self.line, self.col)
    }

    fn run(&mut self) -> Result<Vec<Token>, ParseError> {
        let mut out = Vec::new();
        loop {
            self.skip_trivia();
            let line = self.line;
            let col = self.col;
            let Some(c) = self.peek() else {
                out.push(Token {
                    tok: Tok::Eof,
                    line,
                    col,
                });
                return Ok(out);
            };
            let tok = self.next_token(c)?;
            out.push(Token { tok, line, col });
        }
    }

    /// Skip whitespace and `// ...` line comments. Comment *attachment* to
    /// `$comment` is deferred (spec sections 1 and 12), so comments are simply
    /// discarded here.
    fn skip_trivia(&mut self) {
        loop {
            match self.peek() {
                Some(c) if c.is_whitespace() => {
                    self.bump();
                }
                Some('/') if self.peek2() == Some('/') => {
                    while let Some(c) = self.peek() {
                        if c == '\n' {
                            break;
                        }
                        self.bump();
                    }
                }
                _ => return,
            }
        }
    }

    fn next_token(&mut self, c: char) -> Result<Tok, ParseError> {
        match c {
            '0'..='9' => self.lex_number(),
            '"' => self.lex_string(),
            '`' => self.lex_template(),
            c if is_ident_start(c) => Ok(self.lex_ident()),
            _ => self.lex_symbol(c),
        }
    }

    fn lex_symbol(&mut self, c: char) -> Result<Tok, ParseError> {
        self.bump();
        let n = self.peek();
        let tok = match c {
            '(' => Tok::LParen,
            ')' => Tok::RParen,
            '[' => Tok::LBracket,
            ']' => Tok::RBracket,
            '{' => Tok::LBrace,
            '}' => Tok::RBrace,
            ',' => Tok::Comma,
            ':' => Tok::Colon,
            '.' => {
                if n == Some('.') && self.peek2() == Some('.') {
                    self.bump();
                    self.bump();
                    Tok::DotDotDot
                } else {
                    Tok::Dot
                }
            }
            '+' => {
                if n == Some('+') {
                    self.bump();
                    Tok::PlusPlus
                } else {
                    Tok::Plus
                }
            }
            '-' => {
                if n == Some('>') {
                    self.bump();
                    Tok::Arrow
                } else {
                    Tok::Minus
                }
            }
            '*' => Tok::Star,
            '/' => Tok::Slash,
            '%' => Tok::Percent,
            '!' => {
                if n == Some('=') {
                    self.bump();
                    Tok::BangEq
                } else {
                    Tok::Bang
                }
            }
            '=' => {
                if n == Some('=') {
                    self.bump();
                    Tok::EqEq
                } else if n == Some('>') {
                    self.bump();
                    Tok::FatArrow
                } else {
                    return Err(self.err("unexpected '='; use '==' or '=>'"));
                }
            }
            '<' => {
                if n == Some('=') {
                    self.bump();
                    Tok::LtEq
                } else {
                    Tok::Lt
                }
            }
            '>' => {
                if n == Some('=') {
                    self.bump();
                    Tok::GtEq
                } else {
                    Tok::Gt
                }
            }
            '&' => {
                if n == Some('&') {
                    self.bump();
                    Tok::AndAnd
                } else {
                    Tok::Amp
                }
            }
            '|' => {
                if n == Some('|') {
                    self.bump();
                    Tok::OrOr
                } else {
                    return Err(self.err("unexpected '|'; use '||'"));
                }
            }
            other => return Err(self.err(format!("unexpected character '{other}'"))),
        };
        Ok(tok)
    }

    fn lex_ident(&mut self) -> Tok {
        let mut s = String::new();
        while let Some(c) = self.peek() {
            if is_ident_continue(c) {
                s.push(c);
                self.bump();
            } else {
                break;
            }
        }
        Tok::Ident(s)
    }

    fn lex_number(&mut self) -> Result<Tok, ParseError> {
        let mut s = String::new();
        while let Some(c) = self.peek() {
            if c.is_ascii_digit() {
                s.push(c);
                self.bump();
            } else {
                break;
            }
        }
        let mut is_float = false;
        // Fractional part only if a digit follows the '.', otherwise the '.'
        // is a property-access Dot (e.g. `items[0].name`).
        if self.peek() == Some('.') && self.peek2().is_some_and(|c| c.is_ascii_digit()) {
            is_float = true;
            s.push('.');
            self.bump();
            while let Some(c) = self.peek() {
                if c.is_ascii_digit() {
                    s.push(c);
                    self.bump();
                } else {
                    break;
                }
            }
        }
        if matches!(self.peek(), Some('e' | 'E')) {
            is_float = true;
            s.push('e');
            self.bump();
            if matches!(self.peek(), Some('+' | '-')) {
                s.push(self.peek().unwrap());
                self.bump();
            }
            while let Some(c) = self.peek() {
                if c.is_ascii_digit() {
                    s.push(c);
                    self.bump();
                } else {
                    break;
                }
            }
        }

        let num = if is_float {
            let f: f64 = s
                .parse()
                .map_err(|_| self.err(format!("invalid number literal '{s}'")))?;
            Number::from_f64(f)
                .ok_or_else(|| self.err(format!("number literal '{s}' is not finite")))?
        } else if let Ok(n) = s.parse::<i64>() {
            Number::from(n)
        } else {
            // Integer too large for i64: fall back to f64.
            let f: f64 = s
                .parse()
                .map_err(|_| self.err(format!("invalid number literal '{s}'")))?;
            Number::from_f64(f)
                .ok_or_else(|| self.err(format!("number literal '{s}' is not finite")))?
        };
        Ok(Tok::Num(num))
    }

    fn lex_string(&mut self) -> Result<Tok, ParseError> {
        self.bump(); // opening quote
        let mut s = String::new();
        loop {
            match self.peek() {
                None => return Err(self.err("unterminated string literal")),
                Some('"') => {
                    self.bump();
                    return Ok(Tok::Str(s));
                }
                Some('\\') => {
                    self.bump();
                    let esc = self.read_escape('"')?;
                    s.push_str(&esc);
                }
                Some(c) => {
                    s.push(c);
                    self.bump();
                }
            }
        }
    }

    /// Read a JSON-style escape sequence (the leading backslash already
    /// consumed). `quote` is the closing delimiter for context in errors.
    fn read_escape(&mut self, _quote: char) -> Result<String, ParseError> {
        let Some(c) = self.peek() else {
            return Err(self.err("unterminated escape sequence"));
        };
        self.bump();
        let out = match c {
            '"' => '"'.to_string(),
            '\\' => '\\'.to_string(),
            '/' => '/'.to_string(),
            'b' => '\u{0008}'.to_string(),
            'f' => '\u{000C}'.to_string(),
            'n' => '\n'.to_string(),
            'r' => '\r'.to_string(),
            't' => '\t'.to_string(),
            'u' => {
                let cp = self.read_hex4()?;
                if (0xD800..=0xDBFF).contains(&cp) {
                    // High surrogate: expect a "\uXXXX" low surrogate.
                    if self.peek() == Some('\\') && self.peek2() == Some('u') {
                        self.bump();
                        self.bump();
                        let low = self.read_hex4()?;
                        if (0xDC00..=0xDFFF).contains(&low) {
                            let combined =
                                0x10000 + ((cp - 0xD800) << 10) + (low - 0xDC00);
                            char::from_u32(combined)
                                .unwrap_or('\u{FFFD}')
                                .to_string()
                        } else {
                            return Err(self.err("invalid low surrogate in \\u escape"));
                        }
                    } else {
                        return Err(self.err("unpaired high surrogate in \\u escape"));
                    }
                } else {
                    char::from_u32(cp).unwrap_or('\u{FFFD}').to_string()
                }
            }
            other => return Err(self.err(format!("invalid escape '\\{other}'"))),
        };
        Ok(out)
    }

    fn read_hex4(&mut self) -> Result<u32, ParseError> {
        let mut v: u32 = 0;
        for _ in 0..4 {
            let Some(c) = self.peek() else {
                return Err(self.err("incomplete \\u escape"));
            };
            let d = c
                .to_digit(16)
                .ok_or_else(|| self.err("invalid hex digit in \\u escape"))?;
            v = v * 16 + d;
            self.bump();
        }
        Ok(v)
    }

    /// Lex a backtick template into literal spans + raw hole sources.
    /// Escapes handled inside literal spans: `` \` `` -> backtick, `\$` -> `$`
    /// (so `\${` yields a literal `${`), `\\` -> backslash, plus the standard
    /// JSON escapes.
    fn lex_template(&mut self) -> Result<Tok, ParseError> {
        self.bump(); // opening backtick
        let mut parts: Vec<TemplatePart> = Vec::new();
        let mut cur = String::new();
        loop {
            match self.peek() {
                None => return Err(self.err("unterminated template string")),
                Some('`') => {
                    self.bump();
                    parts.push(TemplatePart::Lit(cur));
                    return Ok(Tok::Template(parts));
                }
                Some('\\') => {
                    self.bump();
                    // `\$` -> literal `$` (covers `\${`); otherwise a JSON escape.
                    if self.peek() == Some('$') {
                        self.bump();
                        cur.push('$');
                    } else {
                        let esc = self.read_escape('`')?;
                        cur.push_str(&esc);
                    }
                }
                Some('$') if self.peek2() == Some('{') => {
                    self.bump();
                    self.bump();
                    parts.push(TemplatePart::Lit(std::mem::take(&mut cur)));
                    let raw = self.read_hole()?;
                    parts.push(TemplatePart::Hole(raw));
                }
                Some(c) => {
                    cur.push(c);
                    self.bump();
                }
            }
        }
    }

    /// Capture the raw source of a `${ ... }` hole (the `${` already consumed),
    /// consuming through the matching `}`. Tracks brace depth and skips over
    /// nested string literals so a `}` inside a string does not close the hole.
    fn read_hole(&mut self) -> Result<String, ParseError> {
        let mut raw = String::new();
        let mut depth = 1usize;
        loop {
            match self.peek() {
                None => return Err(self.err("unterminated ${ ... } template hole")),
                Some('"') => {
                    // Copy the whole string literal verbatim into the raw source.
                    raw.push('"');
                    self.bump();
                    loop {
                        match self.peek() {
                            None => return Err(self.err("unterminated string in template hole")),
                            Some('\\') => {
                                raw.push('\\');
                                self.bump();
                                if let Some(c) = self.peek() {
                                    raw.push(c);
                                    self.bump();
                                }
                            }
                            Some('"') => {
                                raw.push('"');
                                self.bump();
                                break;
                            }
                            Some(c) => {
                                raw.push(c);
                                self.bump();
                            }
                        }
                    }
                }
                Some('{') => {
                    depth += 1;
                    raw.push('{');
                    self.bump();
                }
                Some('}') => {
                    depth -= 1;
                    self.bump();
                    if depth == 0 {
                        return Ok(raw);
                    }
                    raw.push('}');
                }
                Some(c) => {
                    raw.push(c);
                    self.bump();
                }
            }
        }
    }
}

fn is_ident_start(c: char) -> bool {
    c == '_' || c.is_ascii_alphabetic()
}

fn is_ident_continue(c: char) -> bool {
    c == '_' || c.is_ascii_alphanumeric()
}

File: typescript/src/shorthand/lexer.ts
/**
 * Hand-written scanner: `.jfn` source text -> a flat token stream.
 */

import { ParseError } from "./error";

/** One piece of a backtick template: a literal span or a `${ ... }` hole whose
 * raw (undecoded) expression source is parsed later. */
export type TemplatePart = { kind: "lit"; value: string } | { kind: "hole"; value: string };

/** A lexical token. String and number literals carry their decoded value
 * (escapes resolved), so the parser never re-decodes. */
export type Tok =
  | { type: "num"; value: number }
  | { type: "str"; value: string }
  | { type: "ident"; value: string }
  | { type: "template"; parts: TemplatePart[] }
  | { type: TokPunct };

export type TokPunct =
  | "lparen"
  | "rparen"
  | "lbracket"
  | "rbracket"
  | "lbrace"
  | "rbrace"
  | "comma"
  | "colon"
  | "dot"
  | "dotdotdot" // ...  (rest parameter or spread)
  | "arrow" //     ->   (function result annotation or function type)
  | "fatarrow" //  =>   (function literal)
  | "amp" //       &    (function reference)
  | "bang" //      !
  | "plus"
  | "plusplus" //  ++
  | "minus"
  | "star"
  | "slash"
  | "percent"
  | "equals" //    =    (type declaration)
  | "eqeq" //      ==
  | "bangeq" //    !=
  | "lt"
  | "lteq"
  | "gt"
  | "gteq"
  | "andand" //    &&
  | "oror" //      ||
  | "pipe" //      |    (type union)
  | "question" //  ?    (optional key, parameter, or function-type slot)
  | "eof";

/** A token plus its 1-based source position and ending line. */
export type Token = {
  tok: Tok;
  line: number;
  col: number;
  endLine: number;
};

/** Tokenize `src`, returning tokens terminated by a single `eof` token. */
export function lex(src: string): Token[] {
  return new Lexer(src).run();
}

class Lexer {
  private chars: string[];
  private i = 0;
  private line = 1;
  private col = 1;

  constructor(src: string) {
    // Iterate by Unicode code point (not UTF-16 code unit).
    this.chars = Array.from(src);
  }

  private peek(): string | undefined {
    return this.chars[this.i];
  }

  private peek2(): string | undefined {
    return this.chars[this.i + 1];
  }

  private bump(): string | undefined {
    const c = this.chars[this.i];
    if (c === undefined) return undefined;
    this.i++;
    if (c === "\n") {
      this.line++;
      this.col = 1;
    } else {
      this.col++;
    }
    return c;
  }

  private err(message: string): ParseError {
    return new ParseError(message, this.line, this.col);
  }

  run(): Token[] {
    const out: Token[] = [];
    for (;;) {
      this.skipTrivia();
      const line = this.line;
      const col = this.col;
      const c = this.peek();
      if (c === undefined) {
        out.push({ tok: { type: "eof" }, line, col, endLine: line });
        return out;
      }
      const tok = this.nextToken(c);
      out.push({ tok, line, col, endLine: this.line });
    }
  }

  /** Skip whitespace plus line and non-nested block comments. Comment
   * attachment to `$comment` is deferred, so comments are discarded here. */
  private skipTrivia(): void {
    for (;;) {
      const c = this.peek();
      if (c !== undefined && isWhitespace(c)) {
        this.bump();
      } else if (c === "/" && this.peek2() === "/") {
        for (;;) {
          const n = this.peek();
          if (n === undefined || n === "\n") break;
          this.bump();
        }
      } else if (c === "/" && this.peek2() === "*") {
        const line = this.line;
        const col = this.col;
        this.bump();
        this.bump();
        for (;;) {
          const n = this.peek();
          if (n === undefined) {
            throw new ParseError("unterminated block comment", line, col);
          }
          if (n === "*" && this.peek2() === "/") {
            this.bump();
            this.bump();
            break;
          }
          this.bump();
        }
      } else {
        return;
      }
    }
  }

  private nextToken(c: string): Tok {
    if (isDigit(c)) return this.lexNumber();
    if (c === '"') return this.lexString();
    if (c === "`") return this.lexTemplate();
    if (isIdentStart(c)) return this.lexIdent();
    return this.lexSymbol(c);
  }

  private lexSymbol(c: string): Tok {
    this.bump();
    const n = this.peek();
    switch (c) {
      case "(":
        return { type: "lparen" };
      case ")":
        return { type: "rparen" };
      case "[":
        return { type: "lbracket" };
      case "]":
        return { type: "rbracket" };
      case "{":
        return { type: "lbrace" };
      case "}":
        return { type: "rbrace" };
      case ",":
        return { type: "comma" };
      case ":":
        return { type: "colon" };
      case ".":
        if (n === "." && this.peek2() === ".") {
          this.bump();
          this.bump();
          return { type: "dotdotdot" };
        }
        return { type: "dot" };
      case "+":
        if (n === "+") {
          this.bump();
          return { type: "plusplus" };
        }
        return { type: "plus" };
      case "-":
        if (n === ">") {
          this.bump();
          return { type: "arrow" };
        }
        return { type: "minus" };
      case "*":
        return { type: "star" };
      case "/":
        return { type: "slash" };
      case "%":
        return { type: "percent" };
      case "!":
        if (n === "=") {
          this.bump();
          return { type: "bangeq" };
        }
        return { type: "bang" };
      case "=":
        if (n === "=") {
          this.bump();
          return { type: "eqeq" };
        }
        if (n === ">") {
          this.bump();
          return { type: "fatarrow" };
        }
        // A lone `=` is the type-declaration separator (`type Name = …`); the
        // term grammar only ever uses `==`/`=>`.
        return { type: "equals" };
      case "<":
        if (n === "=") {
          this.bump();
          return { type: "lteq" };
        }
        return { type: "lt" };
      case ">":
        if (n === "=") {
          this.bump();
          return { type: "gteq" };
        }
        return { type: "gt" };
      case "&":
        if (n === "&") {
          this.bump();
          return { type: "andand" };
        }
        return { type: "amp" };
      case "|":
        if (n === "|") {
          this.bump();
          return { type: "oror" };
        }
        // A lone `|` is the type-union operator (disjoint from the term grammar,
        // which only ever asks for `||`).
        return { type: "pipe" };
      case "?":
        return { type: "question" };
      default:
        throw this.err(`unexpected character '${c}'`);
    }
  }

  private lexIdent(): Tok {
    let s = "";
    for (;;) {
      const c = this.peek();
      if (!isIdentContinue(c)) break;
      s += c;
      this.bump();
    }
    return { type: "ident", value: s };
  }

  private lexNumber(): Tok {
    let s = "";
    for (;;) {
      const c = this.peek();
      if (!isDigit(c)) break;
      s += c;
      this.bump();
    }
    // Fractional part only if a digit follows the '.', otherwise the '.' is a
    // property-access dot (e.g. `items[0].name`).
    if (this.peek() === "." && isDigit(this.peek2())) {
      s += ".";
      this.bump();
      for (;;) {
        const c = this.peek();
        if (!isDigit(c)) break;
        s += c;
        this.bump();
      }
    }
    if (this.peek() === "e" || this.peek() === "E") {
      s += "e";
      this.bump();
      if (this.peek() === "+" || this.peek() === "-") {
        s += this.peek();
        this.bump();
      }
      for (;;) {
        const c = this.peek();
        if (!isDigit(c)) break;
        s += c;
        this.bump();
      }
    }

    const num = Number(s);
    if (!Number.isFinite(num)) {
      throw this.err(`number literal '${s}' is not finite`);
    }
    return { type: "num", value: num };
  }

  private lexString(): Tok {
    this.bump(); // opening quote
    let s = "";
    for (;;) {
      const c = this.peek();
      if (c === undefined) throw this.err("unterminated string literal");
      if (c === '"') {
        this.bump();
        return { type: "str", value: s };
      }
      if (c === "\\") {
        this.bump();
        s += this.readEscape();
      } else {
        s += c;
        this.bump();
      }
    }
  }

  /** Read a JSON-style escape sequence (the leading backslash already
   * consumed). */
  private readEscape(): string {
    const c = this.peek();
    if (c === undefined) throw this.err("unterminated escape sequence");
    this.bump();
    switch (c) {
      case '"':
        return '"';
      case "\\":
        return "\\";
      case "/":
        return "/";
      case "b":
        return "\u0008";
      case "f":
        return "\u000c";
      case "n":
        return "\n";
      case "r":
        return "\r";
      case "t":
        return "\t";
      case "u": {
        const cp = this.readHex4();
        if (cp >= 0xd800 && cp <= 0xdbff) {
          // High surrogate: expect a "\uXXXX" low surrogate.
          if (this.peek() === "\\" && this.peek2() === "u") {
            this.bump();
            this.bump();
            const low = this.readHex4();
            if (low >= 0xdc00 && low <= 0xdfff) {
              const combined = 0x10000 + ((cp - 0xd800) << 10) + (low - 0xdc00);
              return String.fromCodePoint(combined);
            }
            throw this.err("invalid low surrogate in \\u escape");
          }
          throw this.err("unpaired high surrogate in \\u escape");
        }
        return codePointToString(cp);
      }
      default:
        throw this.err(`invalid escape '\\${c}'`);
    }
  }

  private readHex4(): number {
    let v = 0;
    for (let k = 0; k < 4; k++) {
      const c = this.peek();
      if (c === undefined) throw this.err("incomplete \\u escape");
      const d = hexDigit(c);
      if (d === undefined) throw this.err("invalid hex digit in \\u escape");
      v = v * 16 + d;
      this.bump();
    }
    return v;
  }

  /** Lex a backtick template into literal spans + raw hole sources. Escapes
   * handled inside literal spans: `` \` `` -> backtick, `\$` -> `$` (so `\${`
   * yields a literal `${`), `\\` -> backslash, plus the standard JSON escapes. */
  private lexTemplate(): Tok {
    this.bump(); // opening backtick
    const parts: TemplatePart[] = [];
    let cur = "";
    for (;;) {
      const c = this.peek();
      if (c === undefined) throw this.err("unterminated template string");
      if (c === "`") {
        this.bump();
        parts.push({ kind: "lit", value: cur });
        return { type: "template", parts };
      }
      if (c === "\\") {
        this.bump();
        // `\$` -> literal `$` (covers `\${`); otherwise a JSON escape.
        if (this.peek() === "$") {
          this.bump();
          cur += "$";
        } else {
          cur += this.readEscape();
        }
      } else if (c === "$" && this.peek2() === "{") {
        this.bump();
        this.bump();
        parts.push({ kind: "lit", value: cur });
        cur = "";
        parts.push({ kind: "hole", value: this.readHole() });
      } else {
        cur += c;
        this.bump();
      }
    }
  }

  /** Capture the raw source of a `${ ... }` hole (the `${` already consumed),
   * consuming through the matching `}`. Tracks brace depth and copies strings
   * and comments whole so braces inside them do not close the hole. */
  private readHole(): string {
    let raw = "";
    let depth = 1;
    for (;;) {
      const c = this.peek();
      if (c === undefined) throw this.err("unterminated ${ ... } template hole");
      if (c === '"') {
        // Copy the whole string literal verbatim into the raw source.
        raw += '"';
        this.bump();
        for (;;) {
          const s = this.peek();
          if (s === undefined) throw this.err("unterminated string in template hole");
          if (s === "\\") {
            raw += "\\";
            this.bump();
            const next = this.peek();
            if (next !== undefined) {
              raw += next;
              this.bump();
            }
          } else if (s === '"') {
            raw += '"';
            this.bump();
            break;
          } else {
            raw += s;
            this.bump();
          }
        }
      } else if (c === "/" && this.peek2() === "/") {
        // Preserve the comment for the recursive lexer, but do not interpret a
        // brace inside it as part of the template-hole structure.
        while (this.peek() !== undefined && this.peek() !== "\n") {
          raw += this.bump();
        }
      } else if (c === "/" && this.peek2() === "*") {
        const line = this.line;
        const col = this.col;
        raw += this.bump();
        raw += this.bump();
        for (;;) {
          const n = this.peek();
          if (n === undefined) {
            throw new ParseError("unterminated block comment", line, col);
          }
          raw += this.bump();
          if (n === "*" && this.peek() === "/") {
            raw += this.bump();
            break;
          }
        }
      } else if (c === "{") {
        depth++;
        raw += "{";
        this.bump();
      } else if (c === "}") {
        depth--;
        this.bump();
        if (depth === 0) return raw;
        raw += "}";
      } else {
        raw += c;
        this.bump();
      }
    }
  }
}

function isWhitespace(c: string): boolean {
  return /\s/u.test(c);
}

function isDigit(c: string | undefined): c is string {
  return c !== undefined && c >= "0" && c <= "9";
}

function isAsciiAlpha(c: string): boolean {
  return (c >= "a" && c <= "z") || (c >= "A" && c <= "Z");
}

function isIdentStart(c: string): boolean {
  return c === "_" || isAsciiAlpha(c);
}

function isIdentContinue(c: string | undefined): c is string {
  return c !== undefined && (c === "_" || isAsciiAlpha(c) || isDigit(c));
}

function hexDigit(c: string): number | undefined {
  if (c >= "0" && c <= "9") return c.charCodeAt(0) - 0x30;
  if (c >= "a" && c <= "f") return c.charCodeAt(0) - 0x61 + 10;
  if (c >= "A" && c <= "F") return c.charCodeAt(0) - 0x41 + 10;
  return undefined;
}

/** Map a scalar code point to a string, substituting U+FFFD for values in the
 * surrogate range (which are not valid scalar values). */
function codePointToString(cp: number): string {
  if (cp >= 0xd800 && cp <= 0xdfff) return "\ufffd";
  return String.fromCodePoint(cp);
}
